data_structures/arrays/find_triplets_with_0_sum.py

```diff
@@ -22,3 +22,60 @@ def find_triplets_with_0_sum(nums: list[int]) -> list[list[int]]:
         list(x)
         for x in sorted({abc for abc in combinations(sorted(nums), 3) if not sum(abc)})
     ]
+
+
+def find_triplets_with_0_sum_hashing(arr: list) -> list:
+    """
+    Function for finding the triplets with a given sum in the array using hashing.
+
+    Given a list of integers, return elements a, b, c such that a + b + c = 0.
+
+    Args:
+        nums: list of integers
+    Returns:
+        list of lists of integers where sum(each_list) == 0
+    Examples:
+        >>> find_triplets_with_0_sum_hashing([-1, 0, 1, 2, -1, -4])
+        [[-1, 0, 1], [-1, -1, 2]]
+        >>> find_triplets_with_0_sum_hashing([])
+        []
+        >>> find_triplets_with_0_sum_hashing([0, 0, 0])
+        [[0, 0, 0]]
+        >>> find_triplets_with_0_sum_hashing([1, 2, 3, 0, -1, -2, -3])
+        [[-1, 0, 1], [-3, 1, 2], [-2, 0, 2], [-2, -1, 3], [-3, 0, 3]]
+
+    Time complexity: O(N^2)
+    Auxiliary Space: O(N)
```

Comment:
How do we prove these claims?

Reply:
Time complexity: O(N^2) 
--> This approach has two nested for loops, outer loop runs from start till end of the array length and inner loop from start+1  to end, that's why it has N^2 time complexity.
Auxiliary Space: O(N) 
--> This method Creates a hashmap to store the elements so n extra space has been used.